docs/design_proposals/ko-builder.md

```diff
@@ -0,0 +1,240 @@
+    type ArtifactType struct {
+    	// KoArtifact builds images using [ko](https://github.com/google/ko).
+    	KoArtifact *KoArtifact `yaml:"ko,omitempty" yamltags:"oneOf=artifact"`
+    }
+    ```
+
+Example basic config, this will be sufficient for many users:
+
+```yaml
+apiVersion: skaffold/v2beta14
+kind: Config
+build:
+  artifacts:
+  - image: ko://example.com/helloworld
+    ko: {}
+```
+
+The value of the `image` field is the Go import path of the app entry point,
+[prefixed by `ko://`](https://github.com/google/ko/pull/58).
+
+ko requires setting a
+[`KO_DOCKER_REPO`](https://github.com/google/ko#choose-destination)
+environment variable. This determines where container images are pushed.
+The Skaffold
+[default repo](https://skaffold.dev/docs/environment/image-registries/)
+maps directly to this value.
+
+### Open Questions
+
+1.  Should Skaffold embed ko (as a Go module), or shell out?
+
+    Benefits of embedding:
+
+    - Skaffold can pin the ko version it supports in its `go.mod` file. Users
+      wouldn't raise bugs/issues for incompatible version pairings of Skaffold
+      and ko.
+
+    - Reduce toolchain maintenance toil for users. Skaffold users wouldn't need
+      to synchronize ko versions used by different team members or in their CI
+      build, since the Skaffold version determines the ko version.
+
+    - Portability. Skaffold+ko users only need one tool for their container
+      image building needs: the `skaffold` binary. (Plus the Go distribution,
+      of course.) The current `gcr.io/k8s-skaffold/skaffold` container image
+      could serve as a build and deploy image for CI/CD pipeline steps.
+
+    Embedding ko would require some level of documented behavioural stability
+    guarantees for the most ko interfaces that Skaffold would use, such as
+    [`build.Interface`](https://github.com/google/ko/blob/82cabb40bae577ce3bc016e5939fd85889538e8b/pkg/build/build.go#L24)
+    and
+    [`publish.Interface`](https://github.com/google/ko/blob/82cabb40bae577ce3bc016e5939fd85889538e8b/pkg/publish/publish.go#L24),
+    or others?
+
+    Benefits of shelling out:
+
+    - It's an established pattern used by other Skaffold builders.
+
+    - It would allow Skaffold to support a range of ko versions. On the other
+      hand, these versions would need to be tracked and documented.
+
+    - No need to resolve dependency version differences between Skaffold and
+      ko.
+
+    - If a new ko version provided a significant bug fix, there would be no
+      need to release a new version of Skaffold for this fix.
+
+    Shelling out to ko would require some stability guarantees for the
+    `ko publish` subcommand.
+
+    Suggest embedding as a Go module. __Not Yet Resolved__
+
+2.  Should the ko builder be the default for `skaffold init`, instead of
+    buildpacks, for Go apps, when there's no Dockerfile and no Bazel workspace
+    file?
+
+    Suggest yes, to make Skaffold a compelling choice for Go developers.
```

Comment:
Again getting usage numbers on `ko` would help a lot.  I am not aware how popular ko is since the last time i looked into it.

Reply:
synced offline!